File: internal/domain/invoice.go
package domain

import "time"

// Invoice type constants.
const (
	InvoiceTypeRegular    = "regular"
	InvoiceTypeProforma   = "proforma"
	InvoiceTypeCreditNote = "credit_note"
)

// Invoice relation type constants.
const (
	RelationTypeSettlement = "settlement"
	RelationTypeCreditNote = "credit_note"
)

// Invoice status constants.
const (
	InvoiceStatusDraft     = "draft"
	InvoiceStatusSent      = "sent"
	InvoiceStatusPaid      = "paid"
	InvoiceStatusOverdue   = "overdue"
	InvoiceStatusCancelled = "cancelled"
)

// Invoice represents an issued invoice.
type Invoice struct {
	ID            int64
	SequenceID    int64
	InvoiceNumber string
	Type          string // regular, proforma, credit_note
	Status        string // draft, sent, paid, overdue, cancelled

	IssueDate      time.Time
	DueDate        time.Time
	DeliveryDate   time.Time
	VariableSymbol string
	ConstantSymbol string

	// Customer
	CustomerID int64
	Customer   *Contact

	// Currency
	CurrencyCode string
	ExchangeRate Amount // stored as cents, e.g. 2534 = 25.34 CZK per unit

	// Payment
	PaymentMethod string
	BankAccount   string
	BankCode      string
	IBAN          string
	SWIFT         string

	// Amounts
	SubtotalAmount Amount
	VATAmount      Amount
	TotalAmount    Amount
	PaidAmount     Amount

	// Notes
	Notes         string
	InternalNotes string

	// Related invoice (for credit notes, settlements)
	RelatedInvoiceID *int64
	RelationType     string // "", "settlement", "credit_note"

	// Event timestamps
	SentAt *time.Time
	PaidAt *time.Time

	// Line items
	Items []InvoiceItem

	// Timestamps
	CreatedAt time.Time
	UpdatedAt time.Time
	DeletedAt *time.Time
}

// InvoiceItem represents a single line item on an invoice.
type InvoiceItem struct {
	ID             int64
	InvoiceID      int64
	Description    string
	Quantity       Amount // stored as cents for decimal precision (e.g. 250 = 2.50)
	Unit           string // ks, hod, m2, etc.
	UnitPrice      Amount // price per unit in halere
	VATRatePercent int    // 0, 12, 21
	VATAmount      Amount
	TotalAmount    Amount // including VAT
	SortOrder      int
}

// InvoiceSequence defines a numbering sequence for invoices.
type InvoiceSequence struct {
	ID            int64
	Prefix        string // e.g. "FV"
	NextNumber    int    // next number to assign
	Year          int    // sequence year
	FormatPattern string // e.g. "{prefix}{year}{number:04d}"
}

// CalculateTotals recalculates subtotal, VAT, and total from invoice items.
func (inv *Invoice) CalculateTotals() {
	var subtotal, vat, total Amount
	for i := range inv.Items {
		item := &inv.Items[i]
		// item total before VAT = quantity * unit_price / 100
		// (quantity is in cents, so we divide by 100)
		itemSubtotal := Amount(int64(item.Quantity) * int64(item.UnitPrice) / 100)
		itemVAT := itemSubtotal.Multiply(float64(item.VATRatePercent) / 100.0)
		item.VATAmount = itemVAT
		item.TotalAmount = itemSubtotal.Add(itemVAT)

		subtotal = subtotal.Add(itemSubtotal)
		vat = vat.Add(itemVAT)
		total = total.Add(item.TotalAmount)
	}
	inv.SubtotalAmount = subtotal
	inv.VATAmount = vat
	inv.TotalAmount = total
}

// IsOverdue returns true if the invoice is past due and not yet fully paid.
func (inv *Invoice) IsOverdue() bool {
	if inv.Status == InvoiceStatusPaid || inv.Status == InvoiceStatusCancelled {
		return false
	}
	return time.Now().After(inv.DueDate)
}

// IsPaid returns true if the invoice has been fully paid.
func (inv *Invoice) IsPaid() bool {
	return inv.PaidAmount >= inv.TotalAmount && inv.TotalAmount > 0
}

File: rust/crates/zfaktury-domain/src/invoice.rs
use chrono::{NaiveDate, NaiveDateTime};
use std::fmt;

use crate::amount::Amount;
use crate::contact::Contact;

/// Invoice type enum.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum InvoiceType {
    Regular,
    Proforma,
    CreditNote,
}

impl fmt::Display for InvoiceType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            InvoiceType::Regular => write!(f, "regular"),
            InvoiceType::Proforma => write!(f, "proforma"),
            InvoiceType::CreditNote => write!(f, "credit_note"),
        }
    }
}

/// Invoice status enum.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum InvoiceStatus {
    Draft,
    Sent,
    Paid,
    Overdue,
    Cancelled,
}

impl fmt::Display for InvoiceStatus {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            InvoiceStatus::Draft => write!(f, "draft"),
            InvoiceStatus::Sent => write!(f, "sent"),
            InvoiceStatus::Paid => write!(f, "paid"),
            InvoiceStatus::Overdue => write!(f, "overdue"),
            InvoiceStatus::Cancelled => write!(f, "cancelled"),
        }
    }
}

/// Invoice relation type enum.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RelationType {
    None,
    Settlement,
    CreditNote,
}

impl fmt::Display for RelationType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            RelationType::None => write!(f, ""),
            RelationType::Settlement => write!(f, "settlement"),
            RelationType::CreditNote => write!(f, "credit_note"),
        }
    }
}

/// An issued invoice.
#[derive(Debug, Clone)]
pub struct Invoice {
    pub id: i64,
    pub sequence_id: i64,
    pub invoice_number: String,
    pub invoice_type: InvoiceType,
    pub status: InvoiceStatus,

    pub issue_date: NaiveDate,
    pub due_date: NaiveDate,
    pub delivery_date: NaiveDate,
    pub variable_symbol: String,
    pub constant_symbol: String,

    // Customer
    pub customer_id: i64,
    pub customer: Option<Contact>,

    // Currency
    pub currency_code: String,
    pub exchange_rate: Amount,

    // Payment
    pub payment_method: String,
    pub bank_account: String,
    pub bank_code: String,
    pub iban: String,
    pub swift: String,

    // Amounts
    pub subtotal_amount: Amount,
    pub vat_amount: Amount,
    pub total_amount: Amount,
    pub paid_amount: Amount,

    // Notes
    pub notes: String,
    pub internal_notes: String,

    // Related invoice (for credit notes, settlements)
    pub related_invoice_id: Option<i64>,
    pub relation_type: RelationType,

    // Event timestamps
    pub sent_at: Option<NaiveDateTime>,
    pub paid_at: Option<NaiveDateTime>,

    // Line items
    pub items: Vec<InvoiceItem>,

    // Timestamps
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
    pub deleted_at: Option<NaiveDateTime>,
}

impl Invoice {
    /// Recalculates subtotal, VAT, and total from invoice items.
    pub fn calculate_totals(&mut self) {
        let mut subtotal = Amount::ZERO;
        let mut vat = Amount::ZERO;
        let mut total = Amount::ZERO;

        for item in &mut self.items {
            // item subtotal before VAT = quantity * unit_price / 100
            // (quantity is in cents, so we divide by 100)
            let item_subtotal =
                Amount::from_halere(item.quantity.halere() * item.unit_price.halere() / 100);
            let item_vat = item_subtotal.multiply(item.vat_rate_percent as f64 / 100.0);
            item.vat_amount = item_vat;
            item.total_amount = item_subtotal + item_vat;

            subtotal += item_subtotal;
            vat += item_vat;
            total += item.total_amount;
        }

        self.subtotal_amount = subtotal;
        self.vat_amount = vat;
        self.total_amount = total;
    }

    /// Returns true if the invoice is past due and not yet fully paid.
    pub fn is_overdue(&self, today: NaiveDate) -> bool {
        if self.status == InvoiceStatus::Paid || self.status == InvoiceStatus::Cancelled {
            return false;
        }
        today > self.due_date
    }

    /// Returns true if the invoice has been fully paid.
    pub fn is_paid(&self) -> bool {
        self.paid_amount >= self.total_amount && self.total_amount > Amount::ZERO
    }
}

/// A single line item on an invoice.
#[derive(Debug, Clone)]
pub struct InvoiceItem {
    pub id: i64,
    pub invoice_id: i64,
    pub description: String,
    pub quantity: Amount,
    pub unit: String,
    pub unit_price: Amount,
    pub vat_rate_percent: i32,
    pub vat_amount: Amount,
    pub total_amount: Amount,
    pub sort_order: i32,
}

/// A numbering sequence for invoices.
#[derive(Debug, Clone)]
pub struct InvoiceSequence {
    pub id: i64,
    pub prefix: String,
    pub next_number: i32,
    pub year: i32,
    pub format_pattern: String,
}

/// Filtering options for listing invoices.
#[derive(Debug, Clone, Default)]
pub struct InvoiceFilter {
    pub status: Option<InvoiceStatus>,
    pub invoice_type: Option<InvoiceType>,
    pub customer_id: Option<i64>,
    pub date_from: Option<NaiveDate>,
    pub date_to: Option<NaiveDate>,
    pub search: String,
    pub limit: i32,
    pub offset: i32,
}

/// A recorded change of an invoice's status.
#[derive(Debug, Clone)]
pub struct InvoiceStatusChange {
    pub id: i64,
    pub invoice_id: i64,
    pub old_status: InvoiceStatus,
    pub new_status: InvoiceStatus,
    pub changed_at: NaiveDateTime,
    pub note: String,
}

/// A file attachment linked to an invoice.
#[derive(Debug, Clone)]
pub struct InvoiceDocument {
    pub id: i64,
    pub invoice_id: i64,
    pub filename: String,
    pub content_type: String,
    pub storage_path: String,
    pub size: i64,
    pub created_at: NaiveDateTime,
    pub deleted_at: Option<NaiveDateTime>,
}
